src/app/store/auth/effects/auth.effects.ts

```diff
@@ -0,0 +1,52 @@
+import { Injectable } from '@angular/core';
+import { Router } from '@angular/router';
+
+import { Actions, createEffect, Effect, ofType } from '@ngrx/effects';
+
+import { throwError } from 'rxjs';
+import { catchError, map, switchMap, tap } from 'rxjs/operators';
+
+import { AppRoutePath, ErrorMessage } from '@app/enums';
+import { AuthService } from '@app/services';
+import { loginStart, logoutStart, retrieveInitialAuthData, setAuthData } from '@app/store/auth/actions/auth.actions';
+
+@Injectable()
+export class AuthEffects {
+  constructor(private actions$: Actions, private router: Router, private authService: AuthService) { }
+
+  @Effect()
+  retrieveInitialAuthData$ = this.actions$.pipe(
+    ofType(retrieveInitialAuthData),
+    map(() => setAuthData(this.authService.retrieveInitialAuthData())),
+  );
+
+  loginStart$ = createEffect(() => {
+    let userName = '';
+    let authToken = '';
+
+    return this.actions$.pipe(
+      ofType(loginStart),
+      switchMap(({ login, password }) => {
+        userName = login;
+
+        return this.authService.login({ login, password })
+      }),
+      catchError(({ error }) => {
+        alert(`${ErrorMessage.Login}: ${error}`);
+
+        return throwError(error);
+      }),
+      tap(({ token }) => authToken = token),
+      tap(() => this.router.navigate([ AppRoutePath.Courses ])),
+      map(() => setAuthData({ isAuthenticated: true, userName, authToken })),
+    );
```

Comment:
```typescript

return this.actions$.pipe(
  ofType(loginStart),
  switchMap(({ login, password }) => this.authService.login({ login, password }).pipe(       
    map(() => loginSuccess({ userName: login, authToken })), // authToken itself indicated that user isAuthenticated...
    // tap(() => this.router.navigate([ AppRoutePath.Courses ])), // this redirect can be performed separately on loginSuccess as well as setAuthData
    catchError(({ error }) => {
      alert(`${ErrorMessage.Login}: ${error}`); // this can be handled independently
      return of(loginFail({ error, ...otherNecessaryDataForAction }));
    }),
  )),
);
```

Reply:
Thanks for advice